File: java/lib/src/main/java/com/zourzouvillys/httpsig/Algorithm.java
package com.zourzouvillys.httpsig;

/**
 * Signature algorithms defined in RFC 9421.
 */
public enum Algorithm {

    RSA_PSS_SHA512("rsa-pss-sha512"),
    ECDSA_P256_SHA256("ecdsa-p256-sha256"),
    ED25519("ed25519"),
    HMAC_SHA256("hmac-sha256");

    private final String value;

    Algorithm(String value) {
        this.value = value;
    }

    public String value() {
        return value;
    }

    /**
     * Look up an algorithm by its RFC 9421 string identifier.
     *
     * @throws IllegalArgumentException if the value doesn't match any known algorithm
     */
    public static Algorithm fromValue(String value) {
        for (Algorithm alg : values()) {
            if (alg.value.equals(value)) {
                return alg;
            }
        }
        throw new IllegalArgumentException("unknown algorithm: " + value);
    }
}

File: kotlin/lib/src/main/kotlin/com/zourzouvillys/httpsig/Algorithm.kt
package com.zourzouvillys.httpsig

/**
 * Signature algorithms defined in RFC 9421.
 */
sealed class Algorithm(val value: String) {

    data object RsaPssSha512 : Algorithm("rsa-pss-sha512")
    data object EcdsaP256Sha256 : Algorithm("ecdsa-p256-sha256")
    data object Ed25519 : Algorithm("ed25519")
    data object HmacSha256 : Algorithm("hmac-sha256")

    companion object {
        fun fromValue(value: String): Algorithm = when (value) {
            RsaPssSha512.value -> RsaPssSha512
            EcdsaP256Sha256.value -> EcdsaP256Sha256
            Ed25519.value -> Ed25519
            HmacSha256.value -> HmacSha256
            else -> throw IllegalArgumentException("unknown algorithm: $value")
        }
    }

    override fun toString(): String = value
}
